File: src/lib/tauri/time-registry.ts
/**
 * Time Registry - Tauri Bridge
 * 
 * TypeScript interface to the Rust Time Registry.
 * Tracks changes to entities and edges over time.
 */

import { invoke } from '@tauri-apps/api/core';

// =============================================================================
// Types
// =============================================================================

export interface HistoryEntry {
    entry_id: string;
    entity_id: string;
    action: 'create' | 'update' | 'delete';
    data_json: string;
    timestamp: number;
}

export interface EdgeHistoryEntry {
    entry_id: string;
    source_id: string;
    target_id: string;
    action: 'create' | 'update' | 'delete';
    data_json: string;
    timestamp: number;
}

export interface RecordChangeInput {
    entity_id: string;
    action: 'create' | 'update' | 'delete';
    data: unknown;
}

export interface RecordEdgeChangeInput {
    source_id: string;
    target_id: string;
    action: 'create' | 'update' | 'delete';
    data: unknown;
}

// =============================================================================
// Tauri Commands
// =============================================================================

/**
 * Initialize the time registry schema (called during app startup)
 */
export async function timeRegistryInit(): Promise<void> {
    return invoke('time_registry_init');
}

/**
 * Record an entity change
 */
export async function timeRegistryRecordEntityChange(input: RecordChangeInput): Promise<void> {
    return invoke('time_registry_record_entity_change', { input });
}

/**
 * Get entity change history
 */
export async function timeRegistryGetEntityHistory(entityId: string): Promise<HistoryEntry[]> {
    return invoke('time_registry_get_entity_history', { entityId });
}

/**
 * Record an edge change
 */
export async function timeRegistryRecordEdgeChange(input: RecordEdgeChangeInput): Promise<void> {
    return invoke('time_registry_record_edge_change', { input });
}

/**
 * Get edge change history
 */
export async function timeRegistryGetEdgeHistory(sourceId: string, targetId: string): Promise<EdgeHistoryEntry[]> {
    return invoke('time_registry_get_edge_history', { sourceId, targetId });
}

// =============================================================================
// Facade Class
// =============================================================================

/**
 * Time Registry Facade - mirrors the ITemporalStore interface
 */
export class TauriTimeRegistryFacade {
    private initialized = false;

    async init(): Promise<void> {
        if (this.initialized) return;
        try {
            await timeRegistryInit();
            this.initialized = true;
            console.log('[TauriTimeRegistry] Schema initialized');
        } catch (error) {
            console.error('[TauriTimeRegistry] Init failed:', error);
            throw error;
        }
    }

    async recordChange(entityId: string, action: 'create' | 'update' | 'delete', data: unknown): Promise<void> {
        await this.ensureInit();
        return timeRegistryRecordEntityChange({
            entity_id: entityId,
            action,
            data,
        });
    }

    async getEntityHistory(entityId: string): Promise<HistoryEntry[]> {
        await this.ensureInit();
        return timeRegistryGetEntityHistory(entityId);
    }

    async recordEdgeChange(
        sourceId: string,
        targetId: string,
        action: 'create' | 'update' | 'delete',
        data: unknown
    ): Promise<void> {
        await this.ensureInit();
        return timeRegistryRecordEdgeChange({
            source_id: sourceId,
            target_id: targetId,
            action,
            data,
        });
    }

    async getEdgeHistory(sourceId: string, targetId: string): Promise<EdgeHistoryEntry[]> {
        await this.ensureInit();
        return timeRegistryGetEdgeHistory(sourceId, targetId);
    }

    private async ensureInit(): Promise<void> {
        if (!this.initialized) {
            await this.init();
        }
    }
}

// Singleton instance
export const tauriTimeRegistry = new TauriTimeRegistryFacade();

File: src-tauri/src/api/time_registry.rs
//! Time Registry API - Entity and edge change history

use crate::time_registry::commands as cmd;

// ============================================================================
// API Trait
// ============================================================================

#[taurpc::procedures(path = "time_registry", export_to = "../src/bindings.ts")]
pub trait TimeRegistryApi {
    async fn init() -> Result<(), String>;
    async fn record_entity_change(params: String) -> Result<(), String>;
    async fn get_entity_history(entity_id: String) -> Result<String, String>;
    async fn record_edge_change(params: String) -> Result<(), String>;
    async fn get_edge_history(source_id: String, target_id: String) -> Result<String, String>;
}

// ============================================================================
// Implementation
// ============================================================================

#[derive(Clone, Default)]
pub struct TimeRegistryApiImpl;

#[taurpc::resolvers]
impl TimeRegistryApi for TimeRegistryApiImpl {
    async fn init(self) -> Result<(), String> {
        cmd::time_registry_init()
    }
    
    async fn record_entity_change(self, params: String) -> Result<(), String> {
        let input: cmd::RecordChangeInput = serde_json::from_str(&params)
            .map_err(|e| format!("Failed to parse params: {}", e))?;
        cmd::time_registry_record_entity_change(input)
    }
    
    async fn get_entity_history(self, entity_id: String) -> Result<String, String> {
        let result = cmd::time_registry_get_entity_history(entity_id)?;
        serde_json::to_string(&result).map_err(|e| e.to_string())
    }
    
    async fn record_edge_change(self, params: String) -> Result<(), String> {
        let input: cmd::RecordEdgeChangeInput = serde_json::from_str(&params)
            .map_err(|e| format!("Failed to parse params: {}", e))?;
        cmd::time_registry_record_edge_change(input)
    }
    
    async fn get_edge_history(self, source_id: String, target_id: String) -> Result<String, String> {
        let result = cmd::time_registry_get_edge_history(source_id, target_id)?;
        serde_json::to_string(&result).map_err(|e| e.to_string())
    }
}
